### PR TITLE
implement some new methods on IStorageQueue

### PR DESCRIPTION
**Pull request by christothes**
Implemented some new methods on the StorageQueue interface.

- Task<T> GetAsync<T>()
- ClearAsync()
- Task<IEnumerable<CloudQueueMessage>> PeekAsync
- Task<IEnumerable<T>> PeekAsync<T>

### CODE REVIEW

File: Azure.Data.Wrappers/StorageQueue.cs [hunk 1]
```diff
@@ -177,19 +195,30 @@ public virtual CloudQueue Reference
...
+        /// <summary>
+        /// Save Model to queue, as json
+        /// </summary>
+        /// <param name="message">object</param>
+        /// <returns>Task</returns>
+        public virtual async Task SendAsync<T>(T message)
+        {
+            if (null == message)
             {
                 throw new ArgumentNullException("obj");
```
Comment (pmaroun) [inline Azure.Data.Wrappers/StorageQueue.cs]: throw new ArgumentNullException("message");

File: Azure.Data.Wrappers/Interfaces.cs [hunk 2]
```diff
@@ -422,6 +422,36 @@ public interface IQueueObject
...
+        /// <returns>Message</returns>
+        Task<T> GetAsync<T>();
+
+        /// <summary>
+        /// Get Cloud Queue Message
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="count">number of messages to get. Default is 5</param>
+        /// <returns>Message</returns>
+        Task<IEnumerable<T>> GetManyAsync<T>(int count = 5);
```
Comment (pmaroun) [inline Azure.Data.Wrappers/Interfaces.cs]: Get Cloud Queue Message(s)

File: Azure.Data.Wrappers/Interfaces.cs [hunk 3]
```diff
@@ -422,6 +422,36 @@ public interface IQueueObject
...
         /// <returns>Task</returns>
         Task Send(object obj);
+
+        /// <summary>
+        /// Send a generic Object
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="obj"></param>
+        /// <returns></returns>
+        Task SendAsync<T>(T obj);
```
Comment (pmaroun) [inline Azure.Data.Wrappers/Interfaces.cs]: T obj in the interface and T msg in the implementation

File: Azure.Data.Wrappers/Interfaces.cs [hunk 4]
```diff
@@ -471,6 +514,17 @@ public interface IStorageQueue : IQueue<CloudQueueMessage>, IQueueObject, IAzure
...
     }
+
+    public interface IStorageQueue<T> : IQueue<T>, IAzureStorage, IStorageReference<CloudQueue>, IStorageClient<CloudQueueClient>, IQueueCount
+    {
+        #region Methods
+        /// <summary>
+        /// Get Many Cloud Queue Message
+        /// </summary>
+        /// <returns>Messages</returns>
+        Task<IEnumerable<T>> GetMany(int messageCount = 5);
```
Comment (pmaroun) [inline Azure.Data.Wrappers/Interfaces.cs]: GetManyAsync

File: Azure.Data.Wrappers/Interfaces.cs (hunk 3 above)
Comment (pmaroun) [inline Azure.Data.Wrappers/Interfaces.cs]: I see some operations on these interfaces using the Async prefix and some not.  A little confusing to mix both together on async operations.  I generally use this approach, but would tend to stay away from it in this case as existing async operations don't seem to be following the pattern.  thoughts?

File: Azure.Data.Wrappers/StorageQueue.cs [hunk 5]
```diff
@@ -206,6 +235,36 @@ public virtual CloudQueue Reference
...
             await this.reference.DeleteMessageAsync(message);
         }
+
+        /// <summary>
+        /// Clears the contents of the queue
+        /// </summary>
+        /// <returns></returns>
+        public virtual async Task ClearAsync()
+        {
+            await this.reference.ClearAsync();
```
Comment (puppetcoder) [inline Azure.Data.Wrappers/StorageQueue.cs]: Is reference definitely not null?

Comment (puppetcoder) [reply Azure.Data.Wrappers/StorageQueue.cs]: I suppose so

File: Azure.Data.Wrappers/Interfaces.cs (hunk 3 above)
Comment (christothes) [reply Azure.Data.Wrappers/Interfaces.cs]: Yes, it is unconsistent, but I'm having all new methods follow the convention. Granted it is confusing, but I wanted to do only non-breaking changes.

File: Azure.Data.Wrappers/StorageQueue.cs [hunk 6]
```diff
@@ -177,19 +195,30 @@ public virtual CloudQueue Reference
...
+        }
+
+        /// <summary>
+        /// Save Model to queue, as json
+        /// </summary>
+        /// <param name="message">object</param>
+        /// <returns>Task</returns>
+        public virtual async Task SendAsync<T>(T message)
+        {
+            if (null == message)
```
Comment (puppetcoder) [inline Azure.Data.Wrappers/StorageQueue.cs]: I realize this was already there, but this seems to be doing more than wrapping.  Is that the expected behavior?

File: Azure.Data.Wrappers/StorageQueue.cs [hunk 7]
```diff
@@ -155,6 +156,23 @@ public virtual CloudQueue Reference
...
+            var returned = await this.reference.GetMessageAsync(this.visibilityTimeout, null, null);
+            return JsonConvert.DeserializeObject<T>(returned.AsString);
+        }
+
+        public async Task<IEnumerable<T>> GetManyAsync<T>(int messageCount = 5)
+        {
+            if (0 >= messageCount)
+            {
+                messageCount = 1;
+            }
```
Comment (puppetcoder) [inline Azure.Data.Wrappers/StorageQueue.cs]: Why handle this instead of throwing or letting the called method fail?  This seems like an assumption of use.

Comment (christothes) [reply Azure.Data.Wrappers/StorageQueue.cs]: Fair, but I didn't want to change current behavior

File: Azure.Data.Wrappers/StorageQueue.cs (hunk 6 above)
Comment (christothes) [reply Azure.Data.Wrappers/StorageQueue.cs]: it's doing the right thing is the message is just a CloudQueueMessage and serializing it if not.

File: Azure.Data.Wrappers/Interfaces.cs (hunk 3 above)
Comment (jefking) [reply Azure.Data.Wrappers/Interfaces.cs]: Is the general convention in .NET to have the async suffix (even when there are no non-async methods). One point of this library was to 'force' async.

File: Azure.Data.Wrappers/StorageQueue.cs (hunk 1 above)
Comment (jefking) [reply Azure.Data.Wrappers/StorageQueue.cs]: I would prefer not to pull down the stack. Messages coming off of the bus should not be null.

File: Azure.Data.Wrappers/Interfaces.cs (hunk 3 above)
Comment (pmaroun) [reply Azure.Data.Wrappers/Interfaces.cs]: good point @jefking .  i have referenced this in the past when this comes up.  https://stackoverflow.com/questions/15951774/does-the-use-of-the-async-suffix-in-a-method-name-depend-on-whether-the-async  

one part stuck out (not sure if this is applicable here):  

"there are always exceptions to a guideline. The most notable one in the case of naming would be cases where an entire type’s raison d’etre is to provide async-focused functionality, in which case having Async on every method would be overkill, e.g. the methods on Task itself that produce other Tasks."

**Pull request merged**